frame/cosmwasm/src/lib.rs

```diff
@@ -0,0 +1,1071 @@
+		/// * `code_id` the unique code id generated when the code has been uploaded via [`upload`].
+		/// * `salt` the salt, usually used to instantiate the same contract multiple times.
+		/// * `funds` the assets transferred to the contract prior to calling it's `instantiate`
+		///   export.
+		/// * `gas` the maximum gas to use, the remaining is refunded at the end of the transaction.
+		#[transactional]
+		#[pallet::weight(T::WeightInfo::instantiate().saturating_add(*gas))]
+		pub fn instantiate(
+			origin: OriginFor<T>,
+			code_id: CosmwasmCodeId,
+			salt: ContractSaltOf<T>,
+			admin: Option<AccountIdOf<T>>,
+			label: ContractLabelOf<T>,
+			funds: FundsOf<T>,
+			gas: u64,
+			message: ContractMessageOf<T>,
+		) -> DispatchResultWithPostInfo {
+			let who = ensure_signed(origin)?;
+			let mut shared = Self::do_create_vm_shared(gas, false);
+			let outcome = Self::do_extrinsic_instantiate(
+				&mut shared,
+				who,
+				code_id,
+				&salt,
+				admin,
+				label,
+				funds,
+				message,
+			);
+			Self::refund_gas(outcome, gas, shared.gas.remaining())
+		}
+
+		/// Execute a previously instantiated contract.
+		///
+		/// * Emits an `Executed` event.
+		/// * Possibily emit `Emitted` events.
+		///
+		/// Arguments
+		///
+		/// * `origin` the origin dispatching the extrinsic.
+		/// * `code_id` the unique code id generated when the code has been uploaded via [`upload`].
+		/// * `salt` the salt, usually used to instantiate the same contract multiple times.
+		/// * `funds` the assets transferred to the contract prior to calling it's `instantiate`
+		///   export.
+		/// * `gas` the maximum gas to use, the remaining is refunded at the end of the transaction.
+		#[transactional]
+		#[pallet::weight(T::WeightInfo::execute().saturating_add(*gas))]
+		pub fn execute(
+			origin: OriginFor<T>,
+			contract: AccountIdOf<T>,
+			funds: FundsOf<T>,
+			gas: u64,
+			message: ContractMessageOf<T>,
+		) -> DispatchResultWithPostInfo {
+			let who = ensure_signed(origin)?;
+			let mut shared = Self::do_create_vm_shared(gas, false);
+			let outcome = Self::do_extrinsic_execute(&mut shared, who, contract, funds, message);
+			Self::refund_gas(outcome, gas, shared.gas.remaining())
+		}
+	}
+
+	impl<T: Config> Pallet<T> {
+		/// Deterministic contract address computation, equivalent to solidity CREATE2.
+		pub(crate) fn derive_contract_address(
+			instantiator: &AccountIdOf<T>,
+			code_id: CosmwasmCodeId,
+			salt: &[u8],
+		) -> AccountIdOf<T> {
+			let data: Vec<_> = instantiator
+				.as_ref()
+				.iter()
+				.chain(&code_id.to_le_bytes())
+				.chain(salt)
+				.cloned()
+				.collect();
+			UncheckedFrom::unchecked_from(T::Hashing::hash(&data))
+		}
+
+		/// Deterministic contract trie id generation.
+		pub(crate) fn derive_contract_trie_id(
+			contract: &AccountIdOf<T>,
+			nonce: u64,
+		) -> ContractTrieIdOf<T> {
+			let data: Vec<_> =
+				contract.as_ref().iter().chain(&nonce.to_le_bytes()).cloned().collect();
```

Comment:
nit: hash takes a reference, so I think we can avoid the clone + collect, and instead add the data to the hasher?

Reply:
The `Hasher` here is a trait that has a `hash` function only, operating on `[u8]`. Am I missing something?